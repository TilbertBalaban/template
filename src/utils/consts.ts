export const STAR_WARS_API_URL = 'https://swapi.dev/api/';
export const JSON_PLACEHOLDER_API_URL = 'https://jsonplaceholder.typicode.com/';
