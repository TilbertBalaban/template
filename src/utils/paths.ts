export enum paths {
  Homepage = '/',
  SpyingFrog = 'spying_frog',
  RtkCreateApi = 'rtk_create_api',
  RtkCreateAsyncThunk = 'rtk_create_async_thunk',
  RtkCreateApiPost = 'rtk_create_api_post',
  GraphQLAndApollo = 'graphql_and_apollo',
  ScrollAnimation = 'scroll_animation',
  ReactQuery = 'react_query',
}
