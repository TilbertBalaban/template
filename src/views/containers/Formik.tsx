import * as Yup from 'yup';
import { withFormik, FormikProps, FormikErrors, Form, Field } from 'formik';

// Notice: component rerenders when we type in input.
// TODO: test React Hook Form

const ValidationSchema = Yup.object().shape({
  email: Yup.string()
    .email('Invalid email')
    .min(6, 'Too Short!')
    .required('Required'),
});

interface FormValues {
  email: string;
}

interface CustomProps {
  formHeader: string;
}

const InnerForm = ({
  touched,
  errors,
  isSubmitting,
  formHeader,
}: CustomProps & FormikProps<FormValues>) => {
  return (
    <Form>
      <h2>{formHeader}</h2>
      <Field type="email" name="email" />
      {touched.email && errors.email && <div>{errors.email}</div>}

      <button type="submit" disabled={isSubmitting}>
        Submit
      </button>
    </Form>
  );
};

interface MyFormProps {
  initialEmail?: string;
}

const MyForm = withFormik<MyFormProps & CustomProps, FormValues>({
  // Transform outer props into form values
  mapPropsToValues: (props) => {
    return {
      email: props.initialEmail || '',
    };
  },

  // Add a custom validation function (this can be async too!)
  validate: (values: FormValues) => {
    const errors: FormikErrors<FormValues> = {};
    // add custom validation here (cannot I use Yup?)
    console.log(values);
    return errors;
  },
  validationSchema: ValidationSchema,

  handleSubmit: (values) => {
    console.log('submitting', values);
    return values;
  },
})(InnerForm);

const Formik = () => (
  <div>
    <h1>Formik example</h1>
    <MyForm formHeader="Sign up" initialEmail="test@email.com" />
  </div>
);

export { Formik };
