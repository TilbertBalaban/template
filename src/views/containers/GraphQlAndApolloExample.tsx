import { useEffect, useState } from 'react';
import { useMutation, useQuery } from '@apollo/client';
import { GET_ALL_USERS } from '../../graphql/query';
import { CREATE_USER } from '../../graphql/mutation';

interface IUser {
  id: number;
  username: string;
  age: number;
}

export const GraphQlAndApolloExample = () => {
  const { data, loading, refetch } = useQuery(GET_ALL_USERS);
  const [newUser] = useMutation(CREATE_USER);
  const [users, setUsers] = useState<IUser[]>([]);
  const [username, setUsername] = useState('');
  const [age, setAge] = useState(0);

  useEffect(() => {
    if (!loading) {
      setUsers(data.getAllUsers);
    }
  }, [data]);

  const addUser = () => {
    newUser({
      variables: {
        input: {
          username,
          age,
        },
      },
    }).then(() => {
      setUsername('');
      setAge(0);
    });
  };
  const getAll = () => {
    refetch();
  };

  if (loading) {
    return <h1>Loading...</h1>;
  }

  return (
    <div>
      <form>
        <input
          value={username}
          onChange={(e) => setUsername(e.target.value)}
          type="text"
        />
        <input
          value={age}
          onChange={(e) => setAge(+e.target.value)}
          type="number"
        />
        <div className="btns">
          <button type="button" onClick={addUser}>
            Create
          </button>
          <button type="button" onClick={getAll}>
            Get
          </button>
        </div>
      </form>
      <div>
        {users.map((user) => (
          <div className="user">
            {user.id}. {user.username} {user.age}
          </div>
        ))}
      </div>
    </div>
  );
};
