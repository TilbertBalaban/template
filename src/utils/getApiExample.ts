import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IVehicle } from '../models/IVehicle';
import { STAR_WARS_API_URL } from './consts';

export const getApiExample = createApi({
  reducerPath: 'getApiExample',
  baseQuery: fetchBaseQuery({ baseUrl: STAR_WARS_API_URL }),
  endpoints: (build) => ({
    getVehicles: build.query<IVehicle[], number>({
      query: () => ({
        url: 'vehicles/',
      }),
      transformResponse: (response: {
        results: IVehicle[],
      }) => response.results,
    }),
  }),
});
