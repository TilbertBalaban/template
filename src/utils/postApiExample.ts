import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IPost } from '../models/IPost';
import { JSON_PLACEHOLDER_API_URL } from './consts';

export const postApiExample = createApi({
  reducerPath: 'postApiExample',
  baseQuery: fetchBaseQuery({ baseUrl: JSON_PLACEHOLDER_API_URL }),
  tagTypes: ['post'],
  endpoints: (build) => ({
    getPosts: build.query<IPost[], number>({
      query: () => ({
        url: 'posts/',
      }),
      providesTags: () => ['post'],
    }),
    postPost: build.mutation<IPost, IPost>({
      query: (post) => ({
        url: 'posts/',
        method: 'POST',
        body: post,
      }),
      invalidatesTags: () => ['post'],
    }),
  }),
});
