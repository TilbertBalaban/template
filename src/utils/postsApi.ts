import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IPost } from '../models/IPost';
import { JSON_PLACEHOLDER_API_URL } from './consts';

export const postsApi = createApi({
  reducerPath: 'postsApi',
  baseQuery: fetchBaseQuery({ baseUrl: JSON_PLACEHOLDER_API_URL }),
  endpoints: (build) => ({
    getPosts: build.query<IPost[], number>({
      query: () => ({
        url: 'posts/',
      }),
    }),
    postPost: build.mutation<IPost, IPost>({
      query: (post) => ({
        url: 'posts/',
        method: 'POST',
        body: post,
      }),
    }),
  }),
});
