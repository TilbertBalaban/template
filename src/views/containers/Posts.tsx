import { postApiExample } from '../../utils/api/postApiExample';

export const PostsContainer = () => {
  const [postPost, { isLoading, isSuccess }] =
    postApiExample.usePostPostMutation();

  const { isLoading: isLoadingGetPosts } = postApiExample.useGetPostsQuery(0);

  return (
    <div>
      {isLoadingGetPosts && 'isLoadingGetPosts'}
      {isSuccess && (
        <span data-testid="post-add-success-msg">post is added</span>
      )}
      <button
        onClick={() => {
          postPost({
            title: 'title',
            body: 'body',
            userId: 0,
          });
        }}
        data-testid="add-posts-btn"
        type="button"
        disabled={isLoading}
      >
        {isLoading ? 'loading' : 'add post'}
      </button>
    </div>
  );
};
