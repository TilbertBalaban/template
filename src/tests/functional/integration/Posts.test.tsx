import * as React from 'react';

import { screen } from '@testing-library/react';
import { PostsContainer } from '../../../views/containers/Posts';
import { renderWithProviders } from '../../utils';

test('test', () => {
  renderWithProviders(<PostsContainer />);
  const buttonElement = screen.getByRole('button');
  expect(buttonElement).toBeInTheDocument();
  expect(buttonElement).toMatchSnapshot();
  //   screen.debug()
});
