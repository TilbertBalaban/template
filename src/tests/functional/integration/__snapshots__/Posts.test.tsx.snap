// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 1`] = `
<button
  type="button"
>
  add post
</button>
`;
